README.md
Parse cron string to schedule and generate previous or next schedule item

## Parsers ##

**UnixCronParser** - implementation unix cron, see [https://www.ibm.com/docs/en/db2/11.1?topic=task-unix-cron-format](https://www.ibm.com/docs/en/db2/11.1?topic=task-unix-cron-format)

## Usage: ##

```dart

// import library
import 'package:easy_cron/easy_cron.dart';

// create parser instance
final parser = UnixCronParser();

// parse cron time to CronSchedule
final schedule = parser.parse('* * * * *');

// next time
final nextTime = schedule.next();

// previous time
final prevTime = schedule.prev();
```


CHANGELOG.md
## 0.9.2

* Fix range error in weekdays 

## 0.9.1

* Add unix cron non-standard presets (@yearly, @annually, @monthly, @weekly, @daily, @hourly
* Add tests
* Fix method next

## 0.9.0

* Initial release.


test/unit-tests/src/cron_schedule_test.dart
import 'package:easy_cron/easy_cron.dart';
import 'package:test/test.dart';

String report(DateTime date) {
  final year = date.year;
  final month = date.month.toString().padLeft(2, '0');
  final day = date.day.toString().padLeft(2, '0');
  final minute = date.minute.toString().padLeft(2, '0');
  final second = date.second.toString().padLeft(2, '0');
  final hour = date.hour.toString().padLeft(2, '0');

  return '$year-$month-$day $hour:$minute:$second';
}

CronSchedule schedule(String cron) => UnixCronParser().parse(cron);

String next(String cron, {String? startAt}) {
  var date = DateTime(2022, 4, 4);
  if (startAt != null) {
    date = DateTime.parse(startAt);
  }

  final _startAt = date.toUtc().add(Duration(hours: 5));
  return report(schedule(cron).next(_startAt).time);
}

String prev(String cron, {String? startAt}) {
  var date = DateTime(2022, 4, 4);
  if (startAt != null) {
    date = DateTime.parse(startAt);
  }
  final _startAt = date.toUtc().add(Duration(hours: 5));
  return report(schedule(cron).prev(_startAt).time);
}

void complex(String cron) {
  final startAt = DateTime(2022, 4, 4);
  final next = schedule(cron).next(startAt);
  final prev = next.schedule.prev(next.time);
  final nextAtPrev = prev.schedule.next(prev.time);
  expect(report(next.time), equals(report(nextAtPrev.time)));
}

void main() {
  test('next method', () {
    // next minute
    expect(next('* * * * *'), equals('2022-04-04 00:01:00'));

    expect(
      next('* * * * *', startAt: "2022-12-31 23:59:00"),
      equals('2023-01-01 00:00:00'),
    );

    expect(
      next('* * * * *', startAt: "2022-04-12 16:10:00"),
      equals('2022-04-12 16:11:00'),
    );

    // next day
    expect(
      next('0 0 * * *', startAt: "2022-04-04 00:00:00"),
      equals('2022-04-05 00:00:00'),
    );

    expect(
      next('@daily', startAt: "2022-04-04 00:00:00"),
      equals('2022-04-05 00:00:00'),
    );

    expect(
      next('0 0 * * *', startAt: "2022-04-04 22:55:00"),
      equals('2022-04-05 00:00:00'),
    );

    // next monday
    expect(next('0 0 * * 1'), equals('2022-04-11 00:00:00'));

    // At 00:00 on day-of-month 1 and 15 and on Wednesday
    expect(next('0 0 1,15 * 3'), equals('2022-06-01 00:00:00'));

    expect(
      next('0 0 1,15 * 3', startAt: "2022-04-12 16:40"),
      equals('2022-06-01 00:00:00'),
    );

    expect(
      next('0 0 7,15 * 3', startAt: "2022-04-12 16:40"),
      equals('2022-06-15 00:00:00'),
    );

    expect(
      next('0 0 7,15 * 3', startAt: "2022-06-16 16:40"),
      equals('2022-09-07 00:00:00'),
    );

    expect(next('5 0 * 8 *'), equals('2022-08-01 00:05:00'));

    expect(
      next('0 22 * * 1-5', startAt: "2022-04-12 15:48"),
      equals('2022-04-12 22:00:00'),
    );

    expect(
      next('0 0 31 * *', startAt: "2022-04-12 15:54"),
      equals('2022-05-31 00:00:00'),
    );

    expect(
      next('0 0 31 * *', startAt: "2022-12-31 15:54"),
      equals('2023-01-31 00:00:00'),
    );

    expect(next('23 0-20/2 * * *'), equals('2022-04-04 00:23:00'));

    expect(
      next('0 0 29 2 *', startAt: "2022-04-12 15:54"),
      equals('2024-02-29 00:00:00'),
    );

    expect(
      schedule('0 22 * * SAT-SUN').daysOfWeek,
      equals({DateTime.saturday, DateTime.sunday}),
    );

    expect(
      next('0 22 * * SAT-SUN', startAt: "2022-11-24 16:00"),
      equals('2022-11-26 22:00:00'),
    );

    expect(
      next('0 22 * * 6-7', startAt: "2022-11-26 23:00"),
      equals('2022-11-27 22:00:00'),
    );
  });

  test('prev method', () {
    // previous minute
    expect(prev('* * * * *'), equals('2022-04-03 23:59:00'));

    expect(
      prev('* * * * *', startAt: "2022-04-12 16:18:00"),
      equals('2022-04-12 16:17:00'),
    );

    expect(
      prev('* * * * *', startAt: "2023-01-01 00:00:00"),
      equals('2022-12-31 23:59:00'),
    );

    // previous day
    expect(
      prev('0 0 * * *', startAt: "2022-04-12 16:14:00"),
      equals('2022-04-12 00:00:00'),
    );

    expect(
      prev('@daily', startAt: "2022-04-12 16:14:00"),
      equals('2022-04-12 00:00:00'),
    );

    expect(
      prev('0 0 * * *', startAt: "2023-01-01 00:00:00"),
      equals('2022-12-31 00:00:00'),
    );

    // previous monday
    expect(
      prev('0 0 * * 1', startAt: "2022-04-12 23:45:00"),
      equals('2022-04-11 00:00:00'),
    );

    expect(
      prev('0 0 * * 1', startAt: "2022-04-11 23:45:00"),
      equals('2022-04-11 00:00:00'),
    );

    // previous leap year
    expect(
      prev('0 0 29 2 *', startAt: "2022-04-12 15:54"),
      equals('2020-02-29 00:00:00'),
    );

    expect(
      prev('0 0 29 2 *', startAt: "2024-02-29 00:00"),
      equals('2020-02-29 00:00:00'),
    );

    expect(
      prev('0 0 29 2 *', startAt: "2024-03-01 00:00"),
      equals('2024-02-29 00:00:00'),
    );

    expect(prev('0 0 1,15 * 3'), equals('2021-12-15 00:00:00'));
    expect(prev('5 0 * 8 *'), equals('2021-08-31 00:05:00'));
    expect(prev('0 22 * * 1-5'), equals('2022-04-01 22:00:00'));
    expect(prev('23 0-20/2 * * *'), equals('2022-04-03 20:23:00'));
  });

  test('complex test', () {
    complex('* * * * *');
    complex('0 0 * * 1');
    complex('0 0 1,15 * 3');
    complex('5 0 * AUG *');
    complex('1 * * * *');
    complex('* 1 * * *');
    complex('* * 1 * *');
    complex('* * * jan *');
    complex('* * * * MON');
    complex('0 22 * * 1-5');
    complex('23 0-20/2 * * *');
    complex('0 0,12 1 */2 *');
    complex('0 4 8-14 * *');
    complex('0 0 1,15 * 3');
    complex('15 14 1 * *');
    complex('0 22 * * 1-5');
    complex('23 0-20/2 * * *');
    complex('5 4 * * Sun');
    complex('0 0,12 1 */2 *');
    complex('0 4 8-14 * *');
    complex('0 0 1,15 * 3');
    complex('0 0 1,15 * *');
    complex('0 0 * * 1-5');
    complex('32 18 17,21,29 11 mon-wed');
    complex('0-5,22 0-3,5 7,2-5 1,11,7-9 1-3,5-6');
  });
}


test/unit-tests/src/parsers/unix_cron_parser_test.dart
import 'package:easy_cron/easy_cron.dart';
import 'package:test/test.dart';

void main() {
  final parser = UnixCronParser();

  test('Failure parse', () {
    // invalid cron time
    expect(() => parser.parse(''), throwsA(isA<AssertionError>()));
    // only 5 fields
    expect(() => parser.parse('* * * *'), throwsA(isA<AssertionError>()));
    expect(() => parser.parse('* * * * * *'), throwsA(isA<AssertionError>()));

    // invalid separator
    expect(() {
      return parser.parse(',2 ,3 ,1 ,1 ,5');
    }, throwsA(isA<AssertionError>()));

    // invalid range, syntax error
    expect(() {
      return parser.parse('*-10 *-4 *-2 *-1 *-3');
    }, throwsA(isA<ArgumentError>()));

    // invalid range, first > last
    expect(() {
      return parser.parse('12-10 4-2 2-1 5-2 3-1');
    }, throwsA(isA<AssertionError>()));
  });

  test('Successful parse', () {
    // every minute
    expect(
      parser.parse('* * * * *'),
      CronSchedule(
        minutes: {},
        hours: {},
        daysOfMonth: {},
        months: {},
        daysOfWeek: {},
      ),
    );

    // parse simple field
    expect(
      parser.parse('0 0 1 1 0'),
      CronSchedule(
        minutes: {0},
        hours: {0},
        daysOfMonth: {1},
        months: {1},
        daysOfWeek: {7},
      ),
    );

    // parse range and separator
    expect(
      parser.parse('0-5 0-3 2-5 9-11 4-5'),
      CronSchedule(
        minutes: {0, 1, 2, 3, 4, 5},
        hours: {0, 1, 2, 3},
        daysOfMonth: {2, 3, 4, 5},
        months: {9, 10, 11},
        daysOfWeek: {4, 5},
      ),
    );

    // parse range and separator
    expect(
      parser.parse('0-5,22 0-3,5 7,2-5 1,11,7-9 1-3,5-7'),
      CronSchedule(
        minutes: {0, 1, 2, 3, 4, 5, 22},
        hours: {0, 1, 2, 3, 5},
        daysOfMonth: {2, 3, 4, 5, 7},
        months: {1, 7, 8, 9, 11},
        daysOfWeek: {1, 2, 3, 5, 6, 7},
      ),
    );

    // parse range and separator
    expect(
      parser.parse('*/15 * */15 * *'),
      CronSchedule(
        minutes: {0, 15, 30, 45},
        hours: {},
        daysOfMonth: {1, 16, 31},
        months: {},
        daysOfWeek: {},
      ),
    );

    // parse range and separator
    expect(
      parser.parse('10-18/3 10-15/1 10-14/15 * 2-4/2'),
      CronSchedule(
        minutes: {10, 13, 16},
        hours: {10, 11, 12, 13, 14, 15},
        daysOfMonth: {10},
        months: {},
        daysOfWeek: {2, 4},
      ),
    );
  });
}


lib/src/cron_schedule.dart
import 'package:easy_cron/src/cron_time.dart';
import 'package:equatable/equatable.dart';

enum _DateTimeSegment { year, month, day, hour, minute, second }

class CronSchedule extends Equatable {
  final Set<int>? seconds;
  final Set<int>? minutes;
  final Set<int>? hours;
  final Set<int>? daysOfMonth;
  final Set<int>? months;
  final Set<int>? daysOfWeek;
  final Set<int>? years;

  CronSchedule({
    this.seconds,
    this.minutes,
    this.hours,
    this.daysOfMonth,
    this.months,
    this.daysOfWeek,
    this.years,
  });

  @override
  bool get stringify => true;

  @override
  List<Object?> get props {
    return [
      seconds,
      minutes,
      hours,
      daysOfMonth,
      months,
      daysOfWeek,
      years,
    ];
  }

  // default start time
  DateTime _defaultDateTime([DateTime? startAt]) {
    if (startAt != null) {
      return startAt;
    }

    final now = DateTime.now();
    return DateTime(
      now.year,
      now.month,
      now.day,
      now.hour,
      now.minute,
      now.second,
    );
  }

  Set<int> _defaultSet(int min, int max) {
    final set = List.generate(max - min, (index) => min + index + 1).toSet();
    return {min, ...set};
  }

  // ignore: unused_element
  Set<int> get _seconds {
    return seconds?.isNotEmpty == true ? seconds! : _defaultSet(0, 59);
  }

  Set<int> get _minutes {
    return minutes?.isNotEmpty == true ? minutes! : _defaultSet(0, 59);
  }

  Set<int> get _hours {
    return hours?.isNotEmpty == true ? hours! : _defaultSet(0, 23);
  }

  Set<int> get _daysOfMonth {
    return daysOfMonth?.isNotEmpty == true ? daysOfMonth! : _defaultSet(1, 31);
  }

  Set<int> get _months {
    return months?.isNotEmpty == true ? months! : _defaultSet(1, 12);
  }

  Set<int> get _daysOfWeek {
    return daysOfWeek?.isNotEmpty == true ? daysOfWeek! : _defaultSet(1, 7);
  }

  CronTime prev([DateTime? startAt]) {
    var _date = _defaultDateTime(startAt);

    if (seconds?.isNotEmpty == true) {
      _date = _date.subtract(Duration(seconds: 1));
    } else {
      _date = _date.subtract(Duration(minutes: 1));
    }

    while (true) {
      if (minutes?.isNotEmpty == true &&
          _minutes.contains(_date.minute) == false) {
        _date = _dec(_date, _DateTimeSegment.minute);
        continue;
      }

      if (hours?.isNotEmpty == true && _hours.contains(_date.hour) == false) {
        _date = _dec(_date, _DateTimeSegment.hour);
        continue;
      }

      if (daysOfMonth?.isNotEmpty == true &&
          _daysOfMonth.contains(_date.day) == false) {
        _date = _dec(_date, _DateTimeSegment.day);
        continue;
      }

      if (daysOfWeek?.isNotEmpty == true &&
          _daysOfWeek.contains(_date.weekday) == false) {
        _date = _dec(_date, _DateTimeSegment.day);
        continue;
      }

      if (months?.isNotEmpty == true &&
          _months.contains(_date.month) == false) {
        _date = _dec(_date, _DateTimeSegment.day);
        continue;
      }

      return CronTime(time: _date, schedule: this);
    }
  }

  CronTime next([DateTime? startAt]) {
    var _date = _defaultDateTime(startAt);

    if (seconds?.isNotEmpty == true) {
      _date = _date.add(Duration(seconds: 1));
    } else {
      _date = _date.add(Duration(minutes: 1));
    }

    while (true) {
      if (months?.isNotEmpty == true &&
          _months.contains(_date.month) == false) {
        _date = _inc(_date, _DateTimeSegment.month);
        continue;
      }

      if (daysOfWeek?.isNotEmpty == true &&
          _daysOfWeek.contains(_date.weekday) == false) {
        _date = _inc(_date, _DateTimeSegment.day);
        continue;
      }

      if (daysOfMonth?.isNotEmpty == true &&
          _daysOfMonth.contains(_date.day) == false) {
        _date = _inc(_date, _DateTimeSegment.day);
        continue;
      }

      if (hours?.isNotEmpty == true && _hours.contains(_date.hour) == false) {
        _date = _inc(_date, _DateTimeSegment.hour);
        continue;
      }

      if (minutes?.isNotEmpty == true &&
          _minutes.contains(_date.minute) == false) {
        _date = _inc(_date, _DateTimeSegment.minute);
        continue;
      }

      return CronTime(time: _date, schedule: this);
    }
  }

  DateTime _inc(DateTime date, _DateTimeSegment segment) {
    switch (segment) {
      case _DateTimeSegment.year:
        return DateTime(date.year + 1);
      case _DateTimeSegment.month:
        return DateTime(date.year, date.month + 1);
      case _DateTimeSegment.day:
        return DateTime(
          date.year,
          date.month,
          date.day + 1,
        );
      case _DateTimeSegment.hour:
        return DateTime(
          date.year,
          date.month,
          date.day,
          date.hour + 1,
        );
      case _DateTimeSegment.minute:
        return DateTime(
          date.year,
          date.month,
          date.day,
          date.hour,
          date.minute + 1,
        );
      case _DateTimeSegment.second:
        return DateTime(
          date.year,
          date.month,
          date.day,
          date.hour,
          date.minute,
          date.second + 1,
        );
    }
  }

  DateTime _dec(DateTime date, _DateTimeSegment segment) {
    switch (segment) {
      case _DateTimeSegment.year:
        return DateTime(date.year - 1);
      case _DateTimeSegment.month:
        return date.subtract(Duration(days: 1));
      case _DateTimeSegment.day:
        return date.subtract(Duration(days: 1));
      case _DateTimeSegment.hour:
        return date.subtract(Duration(hours: 1));
      case _DateTimeSegment.minute:
        return date.subtract(Duration(minutes: 1));
      case _DateTimeSegment.second:
        return date.subtract(Duration(seconds: 1));
    }
  }
}


lib/src/parsers/unix_cron_parser.dart
import 'package:easy_cron/easy_cron.dart';

enum _ParserType { range, constraint, step, value }

// parser based on spec https://www.ibm.com/docs/en/db2/11.1?topic=task-unix-cron-format

class UnixCronParser implements CronParserInterface {
  @override
  CronSchedule parse(String cron) {
    final fields = _presets(cron).toLowerCase().trim().split(' ');

    assert(
      fields.length == 5,
      'The cron format must be have five time and date fields separated by at '
      'least one blank',
    );

    final minutes = _parseField(fields[0], 0, 59);
    final hours = _parseField(fields[1], 0, 23);
    final dom = _parseField(fields[2], 1, 31);
    final months = _parseField(_normalize(fields[3], _months), 1, 12);
    final dow = _mapToDartWeekday(
      _parseField(_normalize(fields[4], _weekdays), 0, 7),
    );

    return CronSchedule(
      minutes: minutes,
      hours: hours,
      daysOfMonth: dom,
      months: months,
      daysOfWeek: dow,
    );
  }

  /// non-standard shorthands
  String _presets(String cron) {
    if (cron.contains('@')) {
      final presets = {
        'yearly': '0 0 1 1 *',
        'annually': '0 0 1 1 *',
        'monthly': '0 0 1 * *',
        'weekly': '0 0 * * sun',
        'daily': '0 0 * * *',
        'hourly': '0 */1 * * *',
      };

      final preset = cron.substring(1);
      if (!presets.containsKey(preset)) {
        throw ArgumentError();
      }
      return presets[preset]!;
    }

    return cron;
  }

  String _normalize(String field, Map<String, int> dictionary) {
    var normalized = field;

    for (final key in dictionary.keys) {
      normalized = normalized.replaceAll(key, dictionary[key].toString());
    }

    return normalized;
  }

  Map<String, int> get _weekdays {
    return {
      'mon': 1,
      'tue': 2,
      'wed': 3,
      'thu': 4,
      'fri': 5,
      'sat': 6,
      'sun': 0,
    };
  }

  Map<String, int> get _months {
    return {
      'jan': 1,
      'feb': 2,
      'mar': 3,
      'apr': 4,
      'may': 5,
      'jun': 6,
      'jul': 7,
      'aug': 8,
      'sep': 9,
      'oct': 10,
      'nov': 12,
      'dec': 11,
    };
  }

  Set<int> _mapToDartWeekday(Set<int> dow) {
    return dow.map((e) {
      if (e == 0 || e == 7) {
        return DateTime.sunday;
      }
      return e;
    }).toSet();
  }

  // Parse field one of fields minute, hour etc
  Set<int> _parseField(String field, int min, int max) {
    if (field == '*') {
      return {};
    }

    if (field.contains(',')) {
      final parts = field.split(',');
      if (parts.contains('*')) {
        return {};
      }
      final Set<int> result = {};
      for (var element in parts) {
        assert(element.isNotEmpty);

        result.addAll(_parsePart(element, min, max));
      }

      return result;
    }

    return _parsePart(field, min, max);
  }

  _ParserType _parserType(String part) {
    if (RegExp(r'^\d+\-\d+$').hasMatch(part)) {
      return _ParserType.range;
    }

    if (RegExp(r'^\d+\-\d+\/\d+$').hasMatch(part)) {
      return _ParserType.constraint;
    }

    if (RegExp(r'^\*\/\d+$').hasMatch(part)) {
      return _ParserType.step;
    }

    return _ParserType.value;
  }

  Set<int> _parseValue(String part, int min, int max) {
    try {
      final value = int.parse(part);
      assert(value >= min && value <= max);

      return {value};
    } on FormatException {
      throw ArgumentError('Wrong field value $part');
    }
  }

  Set<int> _parseRange(String part, int min, int max) {
    final ranges = part.split('-');
    int from = int.parse(ranges.first);
    int to = int.parse(ranges.last);

    if (max == 7 && to == 0) {
      to = 7;
    }

    assert(from < to && from >= min && to <= max);

    return List.generate(to - from + 1, (i) => i + from).toSet();
  }

  Set<int> _parseConstraint(String part, int min, int max) {
    final matches = RegExp(r'\d+').allMatches(part);
    assert(matches.length == 3);

    final lower = int.parse(matches.first.group(0)!);
    final higher = int.parse(matches.elementAt(1).group(0)!);
    final step = int.parse(matches.last.group(0)!);
    assert(lower < higher && lower >= min && higher <= max);
    assert(higher >= lower);

    if (lower + step > higher) {
      return {lower};
    }

    final count = ((higher - lower) / step).floor() + 1;

    return List.generate(count, (index) {
      return lower + (index * step);
    }).toSet();
  }

  Set<int> _parseStep(String part, int min, int max) {
    final Set<int> result = {};
    final count = int.parse(part.replaceAll('*/', ''));
    var cur = min;
    while (cur <= max) {
      result.add(cur);
      cur += count;
    }

    return result;
  }

  Set<int> _parsePart(String part, int min, int max) {
    switch (_parserType(part)) {
      case _ParserType.value:
        return _parseValue(part, min, max);
      case _ParserType.range:
        return _parseRange(part, min, max);
      case _ParserType.constraint:
        return _parseConstraint(part, min, max);
      case _ParserType.step:
        return _parseStep(part, min, max);
      default:
        throw ArgumentError();
    }
  }
}
